### PR TITLE
add device asserts in scatter/gather kernels

### PR DESCRIPTION
**Pull request by killeent**
Addresses #1368.

Test Plan:
- Test in existing CuTorch repository, verify that tests still pass
- Test with invalid index, and `CUDA_LAUNCH_BLOCKING=1`, verify device-side assert is triggered

### CODE REVIEW

**Comment by apaszke** [comment]
Can you also add a test in PyTorch?

**Comment by colesbury** [comment]
Will PyTorch tests work? Don't device-side asserts require resetting the device after they're triggered? (That would mean killing the PyTorch process)

**Comment by killeent** [comment]
We don't have tests for scatter/gather in PyTorch, perhaps @apaszke is referring to adding the basic unit tests here, not particularly for the failure case?

File: torch/lib/THC/THCTensorScatterGather.cu [hunk 1]
```diff
@@ -93,6 +93,7 @@ __global__ void THCudaTensor_gatherKernel(
                                                           src, &srcOffset);
 
     IndexType indexValue = (IndexType)index.data[indexOffset] - TH_INDEX_BASE;
+    assert(indexValue < src.sizes[dim]);
```
Comment (ngimel) [inline torch/lib/THC/THCTensorScatterGather.cu]: should you also be asserting indexValue>=0, or is index unsigned here?

Comment (soumith) [inline torch/lib/THC/THCTensorScatterGather.cu]: IndexType is signed long afais, so yes >=0 has to be asserted.

File: test/test_torch.py [hunk 2]
```diff
@@ -1036,8 +1036,8 @@ def compare(t, k, dim, dir):
                        random.randint(1, SIZE),
                        random.randint(1, SIZE))
 
-        for kTries in range(3):
-            for dimTries in range(3):
```
Comment (apaszke) [inline test/test_torch.py]: what's the deal with underscores?

Comment (killeent) [inline test/test_torch.py]: The linter I have recommends using _ before unused variables in loops. But obviously this can be changed.

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
this is now merged into master

**Pull request closed without merging**